{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bc01d59ced195153",
   "metadata": {},
   "source": [
    "# Biolink Model Subsetting\n",
    "This notebook demonstrates how to subset the Biolink Model using the Biolink Model schema and a transformation specification in two different ways: first using a transformation specification defined manually in YAML and loaded from the filesystem, and second, using a subset of classes retrieved using SchemaView dynamically."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "91b3677392f95575",
   "metadata": {},
   "source": [
    "First, we import several LinkML helper packages including SchemaView: https://linkml.io/linkml/developers/schemaview.html \n",
    "SchemaView is a LinkML schema introspection tool that provides helpful functions to retrieve model elements by name.  It also supports navigating ancestors and descendants of any model element, including classes, slots, types, and permissible values in enumerations.  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e3f86775ef7a9dcc",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "from pytest import fixture\n",
    "\n",
    "from linkml_runtime.dumpers import yaml_dumper\n",
    "from linkml_map.datamodel.transformer_model import TransformationSpecification, ClassDerivation, SlotDerivation, CopyDirective\n",
    "from linkml_map.inference.schema_mapper import SchemaMapper\n",
    "from linkml_map.session import Session\n",
    "from linkml_runtime.utils.schemaview import SchemaView\n",
    "from linkml_map.utils.loaders import load_specification\n",
    "from linkml_runtime.utils.formatutils import camelcase, underscore\n",
    "from pprint import pprint\n",
    "\n",
    "REPO_ROOT = Path.cwd().parent.parent"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8d3b8f7bc023dae",
   "metadata": {},
   "source": [
    "SchemaView can accept a variety of imports to initialize, including a Path or string representation of a Path or (as we do in this case), a URL of a raw LinkML schema. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8ca1d183bd948863",
   "metadata": {},
   "outputs": [],
   "source": [
    "schema_url = \"https://raw.githubusercontent.com/biolink/biolink-model/master/biolink-model.yaml\"\n",
    "sv = SchemaView(schema_url)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a91dc68f7d162c6f",
   "metadata": {},
   "source": [
    "### Creating a Transformation Specification Manually in YAML format.\n",
    "In our first example, we develop a \"Transform Specification\" for Biolink Model using Class and Slot Derivations as defined by the linkml-map transformation language.  More about that here: https://linkml.io/linkml-map/#TransformationSpecification/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b9f39d577790a0dd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NamedThing\n",
      "Gene\n",
      "Disease\n",
      "PhenotypicFeature\n",
      "Association\n",
      "GeneToPhenotypicFeatureAssociation\n",
      "\n",
      "id\n",
      "name\n",
      "category\n",
      "symbol\n",
      "subject\n",
      "predicate\n",
      "object\n"
     ]
    }
   ],
   "source": [
    "transform_file = REPO_ROOT / \"tests/input/examples/biolink/transform/biolink-example-profile.transform.yaml\"\n",
    "# Initialize Session and SchemaBuilder\n",
    "session = Session()\n",
    "\n",
    "# Set the source schema in the session\n",
    "session.set_source_schema(sv)\n",
    "\n",
    "tr_spec = load_specification(transform_file)\n",
    "mapper = SchemaMapper()\n",
    "mapper.source_schemaview = sv\n",
    "\n",
    "target_schema_obj = mapper.derive_schema(specification=tr_spec,\n",
    "                                         target_schema_id=\"biolink-profile\",\n",
    "                                        target_schema_name=\"BiolinkProfile\")\n",
    "\n",
    "\n",
    "yaml_dumper.dump(target_schema_obj, \"biolink-profile.yaml\")\n",
    "\n",
    "transformed_sv = SchemaView(\"biolink-profile.yaml\")\n",
    "\n",
    "for class_name in transformed_sv.all_classes():\n",
    "    print(class_name)\n",
    "print()\n",
    "for slot_name in transformed_sv.all_slots():\n",
    "    print(slot_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8ac53bfef2f51db",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exception: range: label type\n"
     ]
    }
   ],
   "source": [
    "!gen-pydantic biolink-profile.yaml"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58dcaf83d20f4dea",
   "metadata": {},
   "source": [
    "Note:\n",
    "* Still in development is the tracing of provenance between a source schema and a destination schema.  Right now there is no provenance.  \n",
    "* If a class, slot, enum, or type is not included at all in the Derivation, it will not be pushed forward to the destination schema.  In development is an option to pull all non-specified components of the source model into the destination model.\n",
    "* You can do transformations on a derivation as well, see: https://linkml.io/linkml-map/#examples/Tutorial/#using-expressions\n",
    "* You can transform data as well as schemas, but this is currently \"beta\" level development.\n",
    "* Custom types are not pulled forward; this is a result of an error at the moment."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48d85f450751e59",
   "metadata": {},
   "source": [
    "### Creating a TransformationSpecification using SchemaView and an existing model.\n",
    "\n",
    "In our second example, we use the Biolink Model directly to derive classes and slots programmatically according to a simple list of \"subset classes\" that we want to extract from the main model in order to produce a subset model according to our specification."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5a7c5a345c5f87d",
   "metadata": {},
   "source": [
    "First, we write a method to extract the classes and slots from Biolink Model using SchemaView"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "87000f75753c5faf",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_biolink_class_derivations(sv, subset_classes) -> dict:\n",
    "    \"\"\"\n",
    "    Function to get Biolink class definitions\n",
    "\n",
    "    :param sv: SchemaView object\n",
    "    :param subset_classes: List of classes to subset\n",
    "    :return: Dictionary of class derivations incl slot derivations\n",
    "    \"\"\"\n",
    "    # Example implementation to fetch class definitions\n",
    "    # This should be replaced with the actual implementation\n",
    "    class_derivations ={}\n",
    "    for class_name in subset_classes:\n",
    "        class_derivation = ClassDerivation(populated_from=class_name,\n",
    "                                           name=camelcase(class_name))\n",
    "        induced_slots = sv.class_induced_slots(class_name)\n",
    "        for slot in induced_slots:\n",
    "            slot_derivation = SlotDerivation(populated_from=slot.name, name=underscore(slot.name))\n",
    "            class_derivation.slot_derivations[underscore(slot.name)] = slot_derivation\n",
    "        class_derivations[camelcase(class_name)] = class_derivation\n",
    "    return class_derivations\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "9f1ca647-fda6-493a-aebe-94ee726a7be2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_biolink_class_derivations_direct(sv, subset_classes) -> dict:\n",
    "    \"\"\"\n",
    "    Function to get Biolink class definitions\n",
    "\n",
    "    :param sv: SchemaView object\n",
    "    :param subset_classes: List of classes to subset\n",
    "    :return: Dictionary of class derivations incl slot derivations\n",
    "    \"\"\"\n",
    "    # Example implementation to fetch class definitions\n",
    "    # This should be replaced with the actual implementation\n",
    "    class_derivations ={}\n",
    "    for class_name in subset_classes:\n",
    "        class_derivation = ClassDerivation(populated_from=class_name,\n",
    "                                           name=camelcase(class_name))\n",
    "        for slot in sv.get_class(class_name).slots:\n",
    "            slot_derivation = SlotDerivation(populated_from=slot, name=underscore(slot))\n",
    "            class_derivation.slot_derivations[underscore(slot)] = slot_derivation\n",
    "        class_derivations[camelcase(class_name)] = class_derivation\n",
    "    return class_derivations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "1193a55025345761",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "class derived:  Gene\n",
      "class derived:  Disease\n",
      "class derived:  CaseToPhenotypicFeatureAssociation\n",
      "class derived:  GeneToDiseaseAssociation\n",
      "class derived:  GeneToPhenotypicFeatureAssociation\n",
      "class derived:  Case\n",
      "class derived:  PhenotypicFeature\n",
      "slot derived:  symbol\n",
      "slot derived:  xref\n",
      "slot derived:  has_biological_sequence\n",
      "slot derived:  id\n",
      "slot derived:  in_taxon\n",
      "slot derived:  in_taxon_label\n",
      "slot derived:  provided_by\n",
      "slot derived:  full_name\n",
      "slot derived:  synonym\n",
      "slot derived:  iri\n",
      "slot derived:  category\n",
      "slot derived:  type\n",
      "slot derived:  name\n",
      "slot derived:  description\n",
      "slot derived:  has_attribute\n",
      "slot derived:  deprecated\n",
      "slot derived:  subject\n",
      "slot derived:  predicate\n",
      "slot derived:  object\n",
      "slot derived:  sex_qualifier\n",
      "slot derived:  disease_context_qualifier\n",
      "slot derived:  subject_specialization_qualifier\n",
      "slot derived:  object_specialization_qualifier\n",
      "slot derived:  anatomical_context_qualifier\n",
      "slot derived:  negated\n",
      "slot derived:  qualifier\n",
      "slot derived:  qualifiers\n",
      "slot derived:  publications\n",
      "slot derived:  has_evidence\n",
      "slot derived:  knowledge_source\n",
      "slot derived:  primary_knowledge_source\n",
      "slot derived:  aggregator_knowledge_source\n",
      "slot derived:  knowledge_level\n",
      "slot derived:  agent_type\n",
      "slot derived:  timepoint\n",
      "slot derived:  original_subject\n",
      "slot derived:  original_predicate\n",
      "slot derived:  original_object\n",
      "slot derived:  subject_category\n",
      "slot derived:  object_category\n",
      "slot derived:  subject_closure\n",
      "slot derived:  object_closure\n",
      "slot derived:  subject_category_closure\n",
      "slot derived:  object_category_closure\n",
      "slot derived:  subject_namespace\n",
      "slot derived:  object_namespace\n",
      "slot derived:  subject_label_closure\n",
      "slot derived:  object_label_closure\n",
      "slot derived:  retrieval_source_ids\n",
      "slot derived:  p_value\n",
      "slot derived:  adjusted_p_value\n",
      "slot derived:  has_supporting_studies\n",
      "slot derived:  has_count\n",
      "slot derived:  has_total\n",
      "slot derived:  has_quotient\n",
      "slot derived:  has_percentage\n",
      "slot derived:  subject_aspect_qualifier\n",
      "slot derived:  subject_direction_qualifier\n",
      "slot derived:  object_aspect_qualifier\n",
      "slot derived:  object_direction_qualifier\n",
      "slot derived:  qualified_predicate\n",
      "slot derived:  frequency_qualifier\n",
      "type copied:  string\n",
      "type copied:  integer\n",
      "type copied:  boolean\n",
      "type copied:  float\n",
      "type copied:  double\n",
      "type copied:  decimal\n",
      "type copied:  time\n",
      "type copied:  date\n",
      "type copied:  datetime\n",
      "type copied:  date_or_datetime\n",
      "type copied:  uriorcurie\n",
      "type copied:  curie\n",
      "type copied:  uri\n",
      "type copied:  ncname\n",
      "type copied:  objectidentifier\n",
      "type copied:  nodeidentifier\n",
      "type copied:  jsonpointer\n",
      "type copied:  jsonpath\n",
      "type copied:  sparqlpath\n",
      "type copied:  chemical formula value\n",
      "type copied:  iri type\n",
      "type copied:  label type\n",
      "type copied:  predicate type\n",
      "type copied:  narrative text\n",
      "type copied:  symbol type\n",
      "type copied:  frequency value\n",
      "type copied:  percentage frequency value\n",
      "type copied:  quotient\n",
      "type copied:  unit\n",
      "type copied:  time type\n",
      "type copied:  biological sequence\n"
     ]
    }
   ],
   "source": [
    "session = Session()\n",
    "\n",
    "# Set the source schema in the session\n",
    "session.set_source_schema(sv)\n",
    "\n",
    "subset_classes = [\n",
    "        \"gene\",\n",
    "        \"disease\",\n",
    "        \"case to phenotypic feature association\",\n",
    "        \"gene to disease association\",\n",
    "        \"gene to phenotypic feature association\",\n",
    "        \"case\",\n",
    "        \"phenotypic feature\",\n",
    "    ]\n",
    "\n",
    "class_derivations = get_biolink_class_derivations_direct(sv, subset_classes)\n",
    "copy_type_directives = {\n",
    "    type_name: CopyDirective(element_name=type_name, copy_all=False)\n",
    "    for type_name, type_def in sv.all_types().items()\n",
    "}\n",
    "\n",
    "ts = TransformationSpecification(class_derivations=class_derivations, copy_directives=copy_type_directives)\n",
    "\n",
    "mapper = SchemaMapper()\n",
    "mapper.source_schemaview = sv\n",
    "\n",
    "target_schema_obj = mapper.derive_schema(\n",
    "    specification=ts, target_schema_id=\"biolink-subset\", target_schema_name=\"BiolinkSubset\"\n",
    ")\n",
    "\n",
    "# ugly bit of hacking to demonstrate end-to-end functionality\n",
    "target_schema_obj.types = sv.all_types()\n",
    "\n",
    "yaml_dumper.dump(target_schema_obj, \"biolink-subset.yaml\")\n",
    "\n",
    "transformed_sv = SchemaView(\"biolink-subset.yaml\")\n",
    "\n",
    "for class_name in transformed_sv.all_classes():\n",
    "    print(\"class derived: \", class_name)\n",
    "for slot_name in transformed_sv.all_slots():\n",
    "    print(\"slot derived: \", slot_name)\n",
    "for type_name in transformed_sv.all_types():\n",
    "    print(\"type copied: \", type_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "f560d74e30c83f1e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "name: BiolinkSubset\n",
      "id: biolink-subset\n",
      "imports:\n",
      "- linkml:types\n",
      "prefixes:\n",
      "  AGRKB:\n",
      "    prefix_prefix: AGRKB\n",
      "    prefix_reference: https://www.alliancegenome.org/\n",
      "  apollo:\n",
      "    prefix_prefix: apollo\n",
      "    prefix_reference: https://github.com/GMOD/Apollo\n",
      "  AspGD:\n",
      "    prefix_prefix: AspGD\n",
      "    prefix_reference: http://www.aspergillusgenome.org/cgi-bin/locus.pl?dbid=\n",
      "  biolink:\n",
      "    prefix_prefix: biolink\n",
      "    prefix_reference: https://w3id.org/biolink/vocab/\n",
      "  bioschemas:\n",
      "    prefix_prefix: bioschemas\n",
      "    prefix_reference: https://bioschemas.org/\n",
      "  linkml:\n",
      "    prefix_prefix: linkml\n",
      "    prefix_reference: https://w3id.org/linkml/\n",
      "  CAID:\n",
      "    prefix_prefix: CAID\n",
      "    prefix_reference: http://reg.clinicalgenome.org/redmine/projects/registry/genboree_registry/by_caid?caid=\n",
      "  CHADO:\n",
      "    prefix_prefix: CHADO\n",
      "    prefix_reference: http://gmod.org/wiki/Chado/\n",
      "  ChemBank:\n",
      "    prefix_prefix: ChemBank\n",
      "    prefix_reference: http://chembank.broadinstitute.org/chemistry/viewMolecule.htm?cbid=\n",
      "  CHEMBL.MECHANISM:\n",
      "    prefix_prefix: CHEMBL.MECHANISM\n",
      "    prefix_reference: https://www.ebi.ac.uk/chembl/mechanism/inspect/\n",
      "  CID:\n",
      "    prefix_prefix: CID\n",
      "    prefix_reference: http://pubchem.ncbi.nlm.nih.gov/compound/\n",
      "  CLINVAR:\n",
      "    prefix_prefix: CLINVAR\n",
      "    prefix_reference: http://identifiers.org/clinvar\n",
      "  COAR_RESOURCE:\n",
      "    prefix_prefix: COAR_RESOURCE\n",
      "    prefix_reference: http://purl.org/coar/resource_type/\n",
      "  COG:\n",
      "    prefix_prefix: COG\n",
      "    prefix_reference: https://www.ncbi.nlm.nih.gov/research/cog-project/\n",
      "  ComplexPortal:\n",
      "    prefix_prefix: ComplexPortal\n",
      "    prefix_reference: https://www.ebi.ac.uk/complexportal/complex/\n",
      "  CPT:\n",
      "    prefix_prefix: CPT\n",
      "    prefix_reference: https://www.ama-assn.org/practice-management/cpt/\n",
      "  CTD.CHEMICAL:\n",
      "    prefix_prefix: CTD.CHEMICAL\n",
      "    prefix_reference: http://ctdbase.org/detail.go?type=chem&acc=\n",
      "  CTD.DISEASE:\n",
      "    prefix_prefix: CTD.DISEASE\n",
      "    prefix_reference: http://ctdbase.org/detail.go?type=disease&db=MESH&acc=\n",
      "  CTD.GENE:\n",
      "    prefix_prefix: CTD.GENE\n",
      "    prefix_reference: http://ctdbase.org/detail.go?type=gene&acc=\n",
      "  CTD:\n",
      "    prefix_prefix: CTD\n",
      "    prefix_reference: http://ctdbase.org/\n",
      "  DGIdb:\n",
      "    prefix_prefix: DGIdb\n",
      "    prefix_reference: https://www.dgidb.org/interaction_types\n",
      "  dcat:\n",
      "    prefix_prefix: dcat\n",
      "    prefix_reference: http://www.w3.org/ns/dcat#\n",
      "  dct:\n",
      "    prefix_prefix: dct\n",
      "    prefix_reference: http://purl.org/dc/terms/\n",
      "  dcid:\n",
      "    prefix_prefix: dcid\n",
      "    prefix_reference: https://datacommons.org/browser/\n",
      "  doi:\n",
      "    prefix_prefix: doi\n",
      "    prefix_reference: https://doi.org/\n",
      "  DOID-PROPERTY:\n",
      "    prefix_prefix: DOID-PROPERTY\n",
      "    prefix_reference: http://purl.obolibrary.org/obo/doid#\n",
      "  DrugCentral:\n",
      "    prefix_prefix: DrugCentral\n",
      "    prefix_reference: http://drugcentral.org/drugcard/\n",
      "  ECTO:\n",
      "    prefix_prefix: ECTO\n",
      "    prefix_reference: http://purl.obolibrary.org/obo/ECTO_\n",
      "  EDAM-DATA:\n",
      "    prefix_prefix: EDAM-DATA\n",
      "    prefix_reference: http://edamontology.org/data_\n",
      "  EDAM-FORMAT:\n",
      "    prefix_prefix: EDAM-FORMAT\n",
      "    prefix_reference: http://edamontology.org/format_\n",
      "  EDAM-OPERATION:\n",
      "    prefix_prefix: EDAM-OPERATION\n",
      "    prefix_reference: http://edamontology.org/operation_\n",
      "  EDAM-TOPIC:\n",
      "    prefix_prefix: EDAM-TOPIC\n",
      "    prefix_reference: http://edamontology.org/topic_\n",
      "  EFO:\n",
      "    prefix_prefix: EFO\n",
      "    prefix_reference: http://www.ebi.ac.uk/efo/EFO_\n",
      "  ExO:\n",
      "    prefix_prefix: ExO\n",
      "    prefix_reference: http://purl.obolibrary.org/obo/ExO_\n",
      "  fabio:\n",
      "    prefix_prefix: fabio\n",
      "    prefix_reference: http://purl.org/spar/fabio/\n",
      "  foaf:\n",
      "    prefix_prefix: foaf\n",
      "    prefix_reference: http://xmlns.com/foaf/0.1/\n",
      "  foodb.compound:\n",
      "    prefix_prefix: foodb.compound\n",
      "    prefix_reference: http://foodb.ca/compounds/\n",
      "  foodb.food:\n",
      "    prefix_prefix: foodb.food\n",
      "    prefix_reference: http://foodb.ca/foods/\n",
      "  FYECO:\n",
      "    prefix_prefix: FYECO\n",
      "    prefix_reference: https://www.pombase.org/term/\n",
      "  FYPO:\n",
      "    prefix_prefix: FYPO\n",
      "    prefix_reference: http://purl.obolibrary.org/obo/FYPO_\n",
      "  gff3:\n",
      "    prefix_prefix: gff3\n",
      "    prefix_reference: https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md#\n",
      "  GOREL:\n",
      "    prefix_prefix: GOREL\n",
      "    prefix_reference: http://purl.obolibrary.org/obo/GOREL_\n",
      "  GOP:\n",
      "    prefix_prefix: GOP\n",
      "    prefix_reference: http://purl.obolibrary.org/obo/go#\n",
      "  gpi:\n",
      "    prefix_prefix: gpi\n",
      "    prefix_reference: https://github.com/geneontology/go-annotation/blob/master/specs/gpad-gpi-2-0.md#\n",
      "  GSID:\n",
      "    prefix_prefix: GSID\n",
      "    prefix_reference: https://scholar.google.com/citations?user=\n",
      "  GTEx:\n",
      "    prefix_prefix: GTEx\n",
      "    prefix_reference: https://www.gtexportal.org/home/gene/\n",
      "  GTOPDB:\n",
      "    prefix_prefix: GTOPDB\n",
      "    prefix_reference: https://www.guidetopharmacology.org/GRAC/LigandDisplayForward?ligandId=\n",
      "  gtpo:\n",
      "    prefix_prefix: gtpo\n",
      "    prefix_reference: https://rdf.guidetopharmacology.org/ns/gtpo#\n",
      "  HANCESTRO:\n",
      "    prefix_prefix: HANCESTRO\n",
      "    prefix_reference: http://www.ebi.ac.uk/ancestro/ancestro_\n",
      "  HCPCS:\n",
      "    prefix_prefix: HCPCS\n",
      "    prefix_reference: http://purl.bioontology.org/ontology/HCPCS/\n",
      "  HsapDv:\n",
      "    prefix_prefix: HsapDv\n",
      "    prefix_reference: http://purl.obolibrary.org/obo/HsapDv_\n",
      "  ICD10:\n",
      "    prefix_prefix: ICD10\n",
      "    prefix_reference: https://icd.codes/icd9cm/\n",
      "  ICD9:\n",
      "    prefix_prefix: ICD9\n",
      "    prefix_reference: http://translator.ncats.nih.gov/ICD9_\n",
      "  interpro:\n",
      "    prefix_prefix: interpro\n",
      "    prefix_reference: https://www.ebi.ac.uk/interpro/entry/\n",
      "  INO:\n",
      "    prefix_prefix: INO\n",
      "    prefix_reference: http://purl.obolibrary.org/obo/INO_\n",
      "  isbn:\n",
      "    prefix_prefix: isbn\n",
      "    prefix_reference: https://www.isbn-international.org/identifier/\n",
      "  isni:\n",
      "    prefix_prefix: isni\n",
      "    prefix_reference: https://isni.org/isni/\n",
      "  issn:\n",
      "    prefix_prefix: issn\n",
      "    prefix_reference: https://portal.issn.org/resource/ISSN/\n",
      "  ncats.drug:\n",
      "    prefix_prefix: ncats.drug\n",
      "    prefix_reference: https://drugs.ncats.io/drug/\n",
      "  KEGG.BRITE:\n",
      "    prefix_prefix: KEGG.BRITE\n",
      "    prefix_reference: 'https://bioregistry.io/kegg.brite:'\n",
      "  KEGG:\n",
      "    prefix_prefix: KEGG\n",
      "    prefix_reference: http://www.kegg.jp/entry/\n",
      "  KEGG.GENES:\n",
      "    prefix_prefix: KEGG.GENES\n",
      "    prefix_reference: 'https://bioregistry.io/kegg.genes:bsu:'\n",
      "  KEGG.PATHWAY:\n",
      "    prefix_prefix: KEGG.PATHWAY\n",
      "    prefix_reference: 'https://bioregistry.io/kegg.pathway:'\n",
      "  KEGG.RCLASS:\n",
      "    prefix_prefix: KEGG.RCLASS\n",
      "    prefix_reference: 'https://www.genome.jp/dbget-bin/www_bget?rc:'\n",
      "  LOINC:\n",
      "    prefix_prefix: LOINC\n",
      "    prefix_reference: http://loinc.org/rdf/\n",
      "  MAXO:\n",
      "    prefix_prefix: MAXO\n",
      "    prefix_reference: http://purl.obolibrary.org/obo/MAXO_\n",
      "  medgen:\n",
      "    prefix_prefix: medgen\n",
      "    prefix_reference: https://www.ncbi.nlm.nih.gov/medgen/\n",
      "  metacyc.reaction:\n",
      "    prefix_prefix: metacyc.reaction\n",
      "    prefix_reference: 'http://identifiers.org/metacyc.reaction:'\n",
      "  METANETX.REACTION:\n",
      "    prefix_prefix: METANETX.REACTION\n",
      "    prefix_reference: https://www.metanetx.org/equa_info/\n",
      "  MESH:\n",
      "    prefix_prefix: MESH\n",
      "    prefix_reference: http://id.nlm.nih.gov/mesh/\n",
      "  MI:\n",
      "    prefix_prefix: MI\n",
      "    prefix_reference: http://purl.obolibrary.org/obo/MI_\n",
      "  mirbase:\n",
      "    prefix_prefix: mirbase\n",
      "    prefix_reference: http://identifiers.org/mirbase\n",
      "  mmmp.biomaps:\n",
      "    prefix_prefix: mmmp.biomaps\n",
      "    prefix_reference: 'https://bioregistry.io/mmmp.biomaps:'\n",
      "  MmusDv:\n",
      "    prefix_prefix: MmusDv\n",
      "    prefix_reference: http://purl.obolibrary.org/obo/MMUSDV_\n",
      "  MSigDB:\n",
      "    prefix_prefix: MSigDB\n",
      "    prefix_reference: https://www.gsea-msigdb.org/gsea/msigdb/\n",
      "  NBO-PROPERTY:\n",
      "    prefix_prefix: NBO-PROPERTY\n",
      "    prefix_reference: http://purl.obolibrary.org/obo/nbo#\n",
      "  ncats.bioplanet:\n",
      "    prefix_prefix: ncats.bioplanet\n",
      "    prefix_reference: https://tripod.nih.gov/bioplanet/detail.jsp?pid=\n",
      "  NCBIGene:\n",
      "    prefix_prefix: NCBIGene\n",
      "    prefix_reference: http://identifiers.org/ncbigene/\n",
      "  NCIT-OBO:\n",
      "    prefix_prefix: NCIT-OBO\n",
      "    prefix_reference: http://purl.obolibrary.org/obo/ncit#\n",
      "  NDDF:\n",
      "    prefix_prefix: NDDF\n",
      "    prefix_reference: http://purl.bioontology.org/ontology/NDDF/\n",
      "  NLMID:\n",
      "    prefix_prefix: NLMID\n",
      "    prefix_reference: https://www.ncbi.nlm.nih.gov/nlmcatalog/?term=\n",
      "  OBAN:\n",
      "    prefix_prefix: OBAN\n",
      "    prefix_reference: http://purl.org/oban/\n",
      "  OMIM.PS:\n",
      "    prefix_prefix: OMIM.PS\n",
      "    prefix_reference: https://www.omim.org/phenotypicSeries/\n",
      "  ORCID:\n",
      "    prefix_prefix: ORCID\n",
      "    prefix_reference: https://orcid.org/\n",
      "  orphanet:\n",
      "    prefix_prefix: orphanet\n",
      "    prefix_reference: http://www.orpha.net/ORDO/Orphanet_\n",
      "  os:\n",
      "    prefix_prefix: os\n",
      "    prefix_reference: https://github.com/cmungall/owlstar/blob/master/owlstar.ttl\n",
      "  PANTHER.FAMILY:\n",
      "    prefix_prefix: PANTHER.FAMILY\n",
      "    prefix_reference: http://www.pantherdb.org/panther/family.do?clsAccession=\n",
      "  PathWhiz:\n",
      "    prefix_prefix: PathWhiz\n",
      "    prefix_reference: http://smpdb.ca/pathways/#\n",
      "  pav:\n",
      "    prefix_prefix: pav\n",
      "    prefix_reference: http://purl.org/pav/\n",
      "  PHARMGKB.DRUG:\n",
      "    prefix_prefix: PHARMGKB.DRUG\n",
      "    prefix_reference: https://www.pharmgkb.org/chemical/\n",
      "  PHARMGKB.DISEASE:\n",
      "    prefix_prefix: PHARMGKB.DISEASE\n",
      "    prefix_reference: https://www.pharmgkb.org/disease/\n",
      "  PHARMGKB.GENE:\n",
      "    prefix_prefix: PHARMGKB.GENE\n",
      "    prefix_reference: https://www.pharmgkb.org/gene/\n",
      "  PHARMGKB.PATHWAYS:\n",
      "    prefix_prefix: PHARMGKB.PATHWAYS\n",
      "    prefix_reference: https://www.pharmgkb.org/pathway/\n",
      "  PHARMGKB.VARIANT:\n",
      "    prefix_prefix: PHARMGKB.VARIANT\n",
      "    prefix_reference: https://www.pharmgkb.org/variant/\n",
      "  PHAROS:\n",
      "    prefix_prefix: PHAROS\n",
      "    prefix_reference: http://pharos.nih.gov\n",
      "  PomBase:\n",
      "    prefix_prefix: PomBase\n",
      "    prefix_reference: https://www.pombase.org/gene/\n",
      "  prov:\n",
      "    prefix_prefix: prov\n",
      "    prefix_reference: http://www.w3.org/ns/prov#\n",
      "  qud:\n",
      "    prefix_prefix: qud\n",
      "    prefix_reference: http://qudt.org/1.1/schema/qudt#\n",
      "  REPODB:\n",
      "    prefix_prefix: REPODB\n",
      "    prefix_reference: http://apps.chiragjpgroup.org/repoDB/\n",
      "  ResearchID:\n",
      "    prefix_prefix: ResearchID\n",
      "    prefix_reference: https://publons.com/researcher/\n",
      "  RO:\n",
      "    prefix_prefix: RO\n",
      "    prefix_reference: http://purl.obolibrary.org/obo/RO_\n",
      "  RXNORM:\n",
      "    prefix_prefix: RXNORM\n",
      "    prefix_reference: http://purl.bioontology.org/ontology/RXNORM/\n",
      "  RXCUI:\n",
      "    prefix_prefix: RXCUI\n",
      "    prefix_reference: https://mor.nlm.nih.gov/RxNav/search?searchBy=RXCUI&searchTerm=\n",
      "  schema:\n",
      "    prefix_prefix: schema\n",
      "    prefix_reference: http://schema.org/\n",
      "  ScopusID:\n",
      "    prefix_prefix: ScopusID\n",
      "    prefix_reference: https://www.scopus.com/authid/detail.uri?authorId=\n",
      "  SEED.REACTION:\n",
      "    prefix_prefix: SEED.REACTION\n",
      "    prefix_reference: https://modelseed.org/biochem/reactions/\n",
      "  SEMMEDDB:\n",
      "    prefix_prefix: SEMMEDDB\n",
      "    prefix_reference: https://skr3.nlm.nih.gov/SemMedDB\n",
      "  SIO:\n",
      "    prefix_prefix: SIO\n",
      "    prefix_reference: http://semanticscience.org/resource/SIO_\n",
      "  SNOMEDCT:\n",
      "    prefix_prefix: SNOMEDCT\n",
      "    prefix_reference: http://snomed.info/id/\n",
      "  SPDI:\n",
      "    prefix_prefix: SPDI\n",
      "    prefix_reference: https://api.ncbi.nlm.nih.gov/variation/v0/spdi/\n",
      "  UBERGRAPH:\n",
      "    prefix_prefix: UBERGRAPH\n",
      "    prefix_reference: http://translator.renci.org/ubergraph-axioms.ofn#\n",
      "  UBERON_CORE:\n",
      "    prefix_prefix: UBERON_CORE\n",
      "    prefix_reference: http://purl.obolibrary.org/obo/uberon/core#\n",
      "  UBERON_NONAMESPACE:\n",
      "    prefix_prefix: UBERON_NONAMESPACE\n",
      "    prefix_reference: http://purl.obolibrary.org/obo/core#\n",
      "  STY:\n",
      "    prefix_prefix: STY\n",
      "    prefix_reference: http://purl.bioontology.org/ontology/STY/\n",
      "  UMLSSG:\n",
      "    prefix_prefix: UMLSSG\n",
      "    prefix_reference: https://lhncbc.nlm.nih.gov/semanticnetwork/download/sg_archive/SemGroups-v04.txt\n",
      "  UniProtKB:\n",
      "    prefix_prefix: UniProtKB\n",
      "    prefix_reference: http://purl.uniprot.org/uniprot/\n",
      "  UNIPROT.ISOFORM:\n",
      "    prefix_prefix: UNIPROT.ISOFORM\n",
      "    prefix_reference: http://purl.uniprot.org/isoforms/\n",
      "  VANDF:\n",
      "    prefix_prefix: VANDF\n",
      "    prefix_reference: https://www.nlm.nih.gov/research/umls/sourcereleasedocs/current/VANDF/\n",
      "  UO-PROPERTY:\n",
      "    prefix_prefix: UO-PROPERTY\n",
      "    prefix_reference: http://purl.obolibrary.org/obo/uo#\n",
      "  VMC:\n",
      "    prefix_prefix: VMC\n",
      "    prefix_reference: https://github.com/ga4gh/vr-spec/\n",
      "  WBls:\n",
      "    prefix_prefix: WBls\n",
      "    prefix_reference: http://purl.obolibrary.org/obo/WBls_\n",
      "  WBbt:\n",
      "    prefix_prefix: WBbt\n",
      "    prefix_reference: http://purl.obolibrary.org/obo/WBbt_\n",
      "  WBVocab:\n",
      "    prefix_prefix: WBVocab\n",
      "    prefix_reference: http://bio2rdf.org/wormbase_vocabulary\n",
      "  WIKIDATA:\n",
      "    prefix_prefix: WIKIDATA\n",
      "    prefix_reference: https://www.wikidata.org/entity/\n",
      "  WIKIDATA_PROPERTY:\n",
      "    prefix_prefix: WIKIDATA_PROPERTY\n",
      "    prefix_reference: https://www.wikidata.org/prop/\n",
      "  wgs:\n",
      "    prefix_prefix: wgs\n",
      "    prefix_reference: http://www.w3.org/2003/01/geo/wgs84_pos\n",
      "  XPO:\n",
      "    prefix_prefix: XPO\n",
      "    prefix_reference: http://purl.obolibrary.org/obo/XPO_\n",
      "  Xenbase:\n",
      "    prefix_prefix: Xenbase\n",
      "    prefix_reference: http://www.xenbase.org/gene/showgene.do?method=display&geneId=\n",
      "  PMC:\n",
      "    prefix_prefix: PMC\n",
      "    prefix_reference: http://europepmc.org/articles/PMC\n",
      "default_prefix: biolink-subset/\n",
      "default_range: string\n",
      "types:\n",
      "  string:\n",
      "    name: string\n",
      "    description: A character string\n",
      "    notes:\n",
      "    - In RDF serializations, a slot with range of string is treated as a literal or\n",
      "      type xsd:string.   If you are authoring schemas in LinkML YAML, the type is\n",
      "      referenced with the lower case \"string\".\n",
      "    from_schema: https://w3id.org/linkml/types\n",
      "    exact_mappings:\n",
      "    - schema:Text\n",
      "    base: str\n",
      "    uri: xsd:string\n",
      "  integer:\n",
      "    name: integer\n",
      "    description: An integer\n",
      "    notes:\n",
      "    - If you are authoring schemas in LinkML YAML, the type is referenced with the\n",
      "      lower case \"integer\".\n",
      "    from_schema: https://w3id.org/linkml/types\n",
      "    exact_mappings:\n",
      "    - schema:Integer\n",
      "    base: int\n",
      "    uri: xsd:integer\n",
      "  boolean:\n",
      "    name: boolean\n",
      "    description: A binary (true or false) value\n",
      "    notes:\n",
      "    - If you are authoring schemas in LinkML YAML, the type is referenced with the\n",
      "      lower case \"boolean\".\n",
      "    from_schema: https://w3id.org/linkml/types\n",
      "    exact_mappings:\n",
      "    - schema:Boolean\n",
      "    base: Bool\n",
      "    uri: xsd:boolean\n",
      "    repr: bool\n",
      "  float:\n",
      "    name: float\n",
      "    description: A real number that conforms to the xsd:float specification\n",
      "    notes:\n",
      "    - If you are authoring schemas in LinkML YAML, the type is referenced with the\n",
      "      lower case \"float\".\n",
      "    from_schema: https://w3id.org/linkml/types\n",
      "    exact_mappings:\n",
      "    - schema:Float\n",
      "    base: float\n",
      "    uri: xsd:float\n",
      "  double:\n",
      "    name: double\n",
      "    description: A real number that conforms to the xsd:double specification\n",
      "    notes:\n",
      "    - If you are authoring schemas in LinkML YAML, the type is referenced with the\n",
      "      lower case \"double\".\n",
      "    from_schema: https://w3id.org/linkml/types\n",
      "    close_mappings:\n",
      "    - schema:Float\n",
      "    base: float\n",
      "    uri: xsd:double\n",
      "  decimal:\n",
      "    name: decimal\n",
      "    description: A real number with arbitrary precision that conforms to the xsd:decimal\n",
      "      specification\n",
      "    notes:\n",
      "    - If you are authoring schemas in LinkML YAML, the type is referenced with the\n",
      "      lower case \"decimal\".\n",
      "    from_schema: https://w3id.org/linkml/types\n",
      "    broad_mappings:\n",
      "    - schema:Number\n",
      "    base: Decimal\n",
      "    uri: xsd:decimal\n",
      "  time:\n",
      "    name: time\n",
      "    description: A time object represents a (local) time of day, independent of any\n",
      "      particular day\n",
      "    notes:\n",
      "    - URI is dateTime because OWL reasoners do not work with straight date or time\n",
      "    - If you are authoring schemas in LinkML YAML, the type is referenced with the\n",
      "      lower case \"time\".\n",
      "    from_schema: https://w3id.org/linkml/types\n",
      "    exact_mappings:\n",
      "    - schema:Time\n",
      "    base: XSDTime\n",
      "    uri: xsd:time\n",
      "    repr: str\n",
      "  date:\n",
      "    name: date\n",
      "    description: a date (year, month and day) in an idealized calendar\n",
      "    notes:\n",
      "    - URI is dateTime because OWL reasoners don't work with straight date or time\n",
      "    - If you are authoring schemas in LinkML YAML, the type is referenced with the\n",
      "      lower case \"date\".\n",
      "    from_schema: https://w3id.org/linkml/types\n",
      "    exact_mappings:\n",
      "    - schema:Date\n",
      "    base: XSDDate\n",
      "    uri: xsd:date\n",
      "    repr: str\n",
      "  datetime:\n",
      "    name: datetime\n",
      "    description: The combination of a date and time\n",
      "    notes:\n",
      "    - If you are authoring schemas in LinkML YAML, the type is referenced with the\n",
      "      lower case \"datetime\".\n",
      "    from_schema: https://w3id.org/linkml/types\n",
      "    exact_mappings:\n",
      "    - schema:DateTime\n",
      "    base: XSDDateTime\n",
      "    uri: xsd:dateTime\n",
      "    repr: str\n",
      "  date_or_datetime:\n",
      "    name: date_or_datetime\n",
      "    description: Either a date or a datetime\n",
      "    notes:\n",
      "    - If you are authoring schemas in LinkML YAML, the type is referenced with the\n",
      "      lower case \"date_or_datetime\".\n",
      "    from_schema: https://w3id.org/linkml/types\n",
      "    base: str\n",
      "    uri: linkml:DateOrDatetime\n",
      "    repr: str\n",
      "  uriorcurie:\n",
      "    name: uriorcurie\n",
      "    description: a URI or a CURIE\n",
      "    notes:\n",
      "    - If you are authoring schemas in LinkML YAML, the type is referenced with the\n",
      "      lower case \"uriorcurie\".\n",
      "    from_schema: https://w3id.org/linkml/types\n",
      "    base: URIorCURIE\n",
      "    uri: xsd:anyURI\n",
      "    repr: str\n",
      "  curie:\n",
      "    name: curie\n",
      "    conforms_to: https://www.w3.org/TR/curie/\n",
      "    description: a compact URI\n",
      "    notes:\n",
      "    - If you are authoring schemas in LinkML YAML, the type is referenced with the\n",
      "      lower case \"curie\".\n",
      "    comments:\n",
      "    - in RDF serializations this MUST be expanded to a URI\n",
      "    - in non-RDF serializations MAY be serialized as the compact representation\n",
      "    from_schema: https://w3id.org/linkml/types\n",
      "    base: Curie\n",
      "    uri: xsd:string\n",
      "    repr: str\n",
      "  uri:\n",
      "    name: uri\n",
      "    conforms_to: https://www.ietf.org/rfc/rfc3987.txt\n",
      "    description: a complete URI\n",
      "    notes:\n",
      "    - If you are authoring schemas in LinkML YAML, the type is referenced with the\n",
      "      lower case \"uri\".\n",
      "    comments:\n",
      "    - in RDF serializations a slot with range of uri is treated as a literal or type\n",
      "      xsd:anyURI unless it is an identifier or a reference to an identifier, in which\n",
      "      case it is translated directly to a node\n",
      "    from_schema: https://w3id.org/linkml/types\n",
      "    close_mappings:\n",
      "    - schema:URL\n",
      "    base: URI\n",
      "    uri: xsd:anyURI\n",
      "    repr: str\n",
      "  ncname:\n",
      "    name: ncname\n",
      "    description: Prefix part of CURIE\n",
      "    notes:\n",
      "    - If you are authoring schemas in LinkML YAML, the type is referenced with the\n",
      "      lower case \"ncname\".\n",
      "    from_schema: https://w3id.org/linkml/types\n",
      "    base: NCName\n",
      "    uri: xsd:string\n",
      "    repr: str\n",
      "  objectidentifier:\n",
      "    name: objectidentifier\n",
      "    description: A URI or CURIE that represents an object in the model.\n",
      "    notes:\n",
      "    - If you are authoring schemas in LinkML YAML, the type is referenced with the\n",
      "      lower case \"objectidentifier\".\n",
      "    comments:\n",
      "    - Used for inheritance and type checking\n",
      "    from_schema: https://w3id.org/linkml/types\n",
      "    base: ElementIdentifier\n",
      "    uri: shex:iri\n",
      "    repr: str\n",
      "  nodeidentifier:\n",
      "    name: nodeidentifier\n",
      "    description: A URI, CURIE or BNODE that represents a node in a model.\n",
      "    notes:\n",
      "    - If you are authoring schemas in LinkML YAML, the type is referenced with the\n",
      "      lower case \"nodeidentifier\".\n",
      "    from_schema: https://w3id.org/linkml/types\n",
      "    base: NodeIdentifier\n",
      "    uri: shex:nonLiteral\n",
      "    repr: str\n",
      "  jsonpointer:\n",
      "    name: jsonpointer\n",
      "    conforms_to: https://datatracker.ietf.org/doc/html/rfc6901\n",
      "    description: A string encoding a JSON Pointer. The value of the string MUST conform\n",
      "      to JSON Point syntax and SHOULD dereference to a valid object within the current\n",
      "      instance document when encoded in tree form.\n",
      "    notes:\n",
      "    - If you are authoring schemas in LinkML YAML, the type is referenced with the\n",
      "      lower case \"jsonpointer\".\n",
      "    from_schema: https://w3id.org/linkml/types\n",
      "    base: str\n",
      "    uri: xsd:string\n",
      "    repr: str\n",
      "  jsonpath:\n",
      "    name: jsonpath\n",
      "    conforms_to: https://www.ietf.org/archive/id/draft-goessner-dispatch-jsonpath-00.html\n",
      "    description: A string encoding a JSON Path. The value of the string MUST conform\n",
      "      to JSON Point syntax and SHOULD dereference to zero or more valid objects within\n",
      "      the current instance document when encoded in tree form.\n",
      "    notes:\n",
      "    - If you are authoring schemas in LinkML YAML, the type is referenced with the\n",
      "      lower case \"jsonpath\".\n",
      "    from_schema: https://w3id.org/linkml/types\n",
      "    base: str\n",
      "    uri: xsd:string\n",
      "    repr: str\n",
      "  sparqlpath:\n",
      "    name: sparqlpath\n",
      "    conforms_to: https://www.w3.org/TR/sparql11-query/#propertypaths\n",
      "    description: A string encoding a SPARQL Property Path. The value of the string\n",
      "      MUST conform to SPARQL syntax and SHOULD dereference to zero or more valid objects\n",
      "      within the current instance document when encoded as RDF.\n",
      "    notes:\n",
      "    - If you are authoring schemas in LinkML YAML, the type is referenced with the\n",
      "      lower case \"sparqlpath\".\n",
      "    from_schema: https://w3id.org/linkml/types\n",
      "    base: str\n",
      "    uri: xsd:string\n",
      "    repr: str\n",
      "  chemical formula value:\n",
      "    name: chemical formula value\n",
      "    description: A chemical formula\n",
      "    notes:\n",
      "    - Should be implemented as a stronger type\n",
      "    from_schema: https://w3id.org/biolink/biolink-model\n",
      "    base: str\n",
      "    uri: xsd:string\n",
      "  iri type:\n",
      "    name: iri type\n",
      "    description: An IRI\n",
      "    from_schema: https://w3id.org/biolink/biolink-model\n",
      "    typeof: uriorcurie\n",
      "    uri: xsd:string\n",
      "  label type:\n",
      "    name: label type\n",
      "    description: A string that provides a human-readable name for an entity\n",
      "    from_schema: https://w3id.org/biolink/biolink-model\n",
      "    typeof: string\n",
      "    uri: xsd:string\n",
      "  predicate type:\n",
      "    name: predicate type\n",
      "    description: A CURIE from the biolink related_to hierarchy. For example, biolink:related_to,\n",
      "      biolink:causes, biolink:treats.\n",
      "    from_schema: https://w3id.org/biolink/biolink-model\n",
      "    typeof: uriorcurie\n",
      "    uri: xsd:string\n",
      "  narrative text:\n",
      "    name: narrative text\n",
      "    description: A string that provides a human-readable description of something\n",
      "    from_schema: https://w3id.org/biolink/biolink-model\n",
      "    typeof: string\n",
      "    uri: xsd:string\n",
      "  symbol type:\n",
      "    name: symbol type\n",
      "    from_schema: https://w3id.org/biolink/biolink-model\n",
      "    typeof: string\n",
      "    uri: xsd:string\n",
      "  frequency value:\n",
      "    name: frequency value\n",
      "    from_schema: https://w3id.org/biolink/biolink-model\n",
      "    typeof: string\n",
      "    uri: UO:0000105\n",
      "  percentage frequency value:\n",
      "    name: percentage frequency value\n",
      "    from_schema: https://w3id.org/biolink/biolink-model\n",
      "    typeof: double\n",
      "    uri: UO:0000187\n",
      "  quotient:\n",
      "    name: quotient\n",
      "    from_schema: https://w3id.org/biolink/biolink-model\n",
      "    aliases:\n",
      "    - ratio\n",
      "    typeof: double\n",
      "    uri: UO:0010006\n",
      "  unit:\n",
      "    name: unit\n",
      "    id_prefixes:\n",
      "    - UO\n",
      "    from_schema: https://w3id.org/biolink/biolink-model\n",
      "    exact_mappings:\n",
      "    - qud:Unit\n",
      "    typeof: string\n",
      "    uri: UO:0000000\n",
      "  time type:\n",
      "    name: time type\n",
      "    from_schema: https://w3id.org/biolink/biolink-model\n",
      "    typeof: time\n",
      "    uri: xsd:time\n",
      "  biological sequence:\n",
      "    name: biological sequence\n",
      "    from_schema: https://w3id.org/biolink/biolink-model\n",
      "    typeof: string\n",
      "    uri: xsd:string\n",
      "classes:\n",
      "  Gene:\n",
      "    name: Gene\n",
      "    id_prefixes:\n",
      "    - NCBIGene\n",
      "    - ENSEMBL\n",
      "    - HGNC\n",
      "    - MGI\n",
      "    - ZFIN\n",
      "    - dictyBase\n",
      "    - WB\n",
      "    - WormBase\n",
      "    - FB\n",
      "    - RGD\n",
      "    - SGD\n",
      "    - PomBase\n",
      "    - OMIM\n",
      "    - KEGG.GENES\n",
      "    - UMLS\n",
      "    - Xenbase\n",
      "    - AspGD\n",
      "    - PHARMGKB.GENE\n",
      "    description: A region (or regions) that includes all of the sequence elements\n",
      "      necessary to encode a functional transcript. A gene locus may include regulatory\n",
      "      regions, transcribed regions and/or other functional sequence regions.\n",
      "    in_subset:\n",
      "    - translator_minimal\n",
      "    - model_organism_database\n",
      "    exact_mappings:\n",
      "    - SO:0000704\n",
      "    - SIO:010035\n",
      "    - WIKIDATA:Q7187\n",
      "    - dcid:Gene\n",
      "    narrow_mappings:\n",
      "    - bioschemas:gene\n",
      "    broad_mappings:\n",
      "    - NCIT:C45822\n",
      "    attributes:\n",
      "      symbol:\n",
      "        name: symbol\n",
      "        description: Symbol for a particular thing\n",
      "        exact_mappings:\n",
      "        - AGRKB:symbol\n",
      "        - gpi:DB_Object_Symbol\n",
      "        is_a: node property\n",
      "        domain: named thing\n",
      "        domain_of:\n",
      "        - gene\n",
      "      xref:\n",
      "        name: xref\n",
      "        description: A database cross reference or alternative identifier for a NamedThing\n",
      "          or edge between two NamedThings.  This property should point to a database\n",
      "          record or webpage that supports the existence of the edge, or gives more\n",
      "          detail about the edge. This property can be used on a node or edge to provide\n",
      "          multiple URIs or CURIE cross references.\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        aliases:\n",
      "        - dbxref\n",
      "        - Dbxref\n",
      "        - DbXref\n",
      "        - record_url\n",
      "        - source_record_urls\n",
      "        narrow_mappings:\n",
      "        - gff3:Dbxref\n",
      "        - gpi:DB_Xrefs\n",
      "        domain: named thing\n",
      "        domain_of:\n",
      "        - named thing\n",
      "        - publication\n",
      "        - retrieval source\n",
      "        - gene\n",
      "        - gene product mixin\n",
      "        range: uriorcurie\n",
      "        multivalued: true\n",
      "  Disease:\n",
      "    name: Disease\n",
      "    id_prefixes:\n",
      "    - MONDO\n",
      "    - DOID\n",
      "    - OMIM\n",
      "    - OMIM.PS\n",
      "    - orphanet\n",
      "    - EFO\n",
      "    - UMLS\n",
      "    - MESH\n",
      "    - MEDDRA\n",
      "    - NCIT\n",
      "    - SNOMEDCT\n",
      "    - medgen\n",
      "    - ICD10\n",
      "    - ICD9\n",
      "    - KEGG.DISEASE\n",
      "    - HP\n",
      "    - MP\n",
      "    - PHARMGKB.DISEASE\n",
      "    description: A disorder of structure or function, especially one that produces\n",
      "      specific signs, phenotypes or symptoms or that affects a specific location and\n",
      "      is not simply a direct result of physical injury.  A disposition to undergo\n",
      "      pathological processes that exists in an organism because of one or more disorders\n",
      "      in that organism.\n",
      "    in_subset:\n",
      "    - model_organism_database\n",
      "    - translator_minimal\n",
      "    aliases:\n",
      "    - condition\n",
      "    - disorder\n",
      "    - medical condition\n",
      "    exact_mappings:\n",
      "    - MONDO:0000001\n",
      "    - DOID:4\n",
      "    - NCIT:C2991\n",
      "    - WIKIDATA:Q12136\n",
      "    - SIO:010299\n",
      "    - UMLSSG:DISO\n",
      "    - STY:T047\n",
      "    - dcid:Disease\n",
      "    narrow_mappings:\n",
      "    - STY:T019\n",
      "    - STY:T020\n",
      "    - STY:T048\n",
      "    - STY:T049\n",
      "    - STY:T191\n",
      "    - MONDO:0042489\n",
      "  CaseToPhenotypicFeatureAssociation:\n",
      "    name: CaseToPhenotypicFeatureAssociation\n",
      "    description: An association between a case (e.g. individual patient) and a phenotypic\n",
      "      feature in which the individual has or has had the phenotype.\n",
      "    defining_slots:\n",
      "    - subject\n",
      "    - object\n",
      "  GeneToDiseaseAssociation:\n",
      "    name: GeneToDiseaseAssociation\n",
      "    comments:\n",
      "    - NCIT:R176 refers to the inverse relationship\n",
      "    exact_mappings:\n",
      "    - SIO:000983\n",
      "    close_mappings:\n",
      "    - dcid:DiseaseGeneAssociation\n",
      "    defining_slots:\n",
      "    - subject\n",
      "    - object\n",
      "  GeneToPhenotypicFeatureAssociation:\n",
      "    name: GeneToPhenotypicFeatureAssociation\n",
      "    exact_mappings:\n",
      "    - WBVocab:Gene-Phenotype-Association\n",
      "    defining_slots:\n",
      "    - subject\n",
      "    - object\n",
      "  Case:\n",
      "    name: Case\n",
      "    description: An individual (human) organism that has a patient role in some clinical\n",
      "      context.\n",
      "    aliases:\n",
      "    - patient\n",
      "    - proband\n",
      "  PhenotypicFeature:\n",
      "    name: PhenotypicFeature\n",
      "    id_prefixes:\n",
      "    - HP\n",
      "    - EFO\n",
      "    - NCIT\n",
      "    - UMLS\n",
      "    - MEDDRA\n",
      "    - MP\n",
      "    - ZP\n",
      "    - UPHENO\n",
      "    - APO\n",
      "    - FBcv\n",
      "    - WBPhenotype\n",
      "    - SNOMEDCT\n",
      "    - MESH\n",
      "    - XPO\n",
      "    - FYPO\n",
      "    - TO\n",
      "    description: A combination of entity and quality that makes up a phenotyping statement.\n",
      "      An observable characteristic of an individual resulting from the interaction\n",
      "      of its genotype with its molecular and physical environment.\n",
      "    examples:\n",
      "    - value: MP:0001262\n",
      "      description: decreased body weight\n",
      "    in_subset:\n",
      "    - model_organism_database\n",
      "    aliases:\n",
      "    - sign\n",
      "    - symptom\n",
      "    - phenotype\n",
      "    - trait\n",
      "    - endophenotype\n",
      "    exact_mappings:\n",
      "    - UPHENO:0001001\n",
      "    - SIO:010056\n",
      "    - WIKIDATA:Q104053\n",
      "    - UMLS:C4021819\n",
      "    - NCIT:C16977\n",
      "    - SNOMEDCT:8116006\n",
      "    - MESH:D010641\n",
      "    narrow_mappings:\n",
      "    - STY:T184\n",
      "    - WIKIDATA:Q169872\n",
      "    - WIKIDATA:Q25203551\n",
      "    - ZP:00000000\n",
      "    - FBcv:0001347\n",
      "    - HP:0000118\n",
      "    - MP:0000001\n",
      "    - WBPhenotype:0000886\n",
      "    - XPO:00000000\n",
      "    - FYPO:0000001\n",
      "    - APO:0000017\n",
      "    - TO:0000387\n",
      "    - STY:T190\n",
      "    broad_mappings:\n",
      "    - BFO:0000019\n",
      "    - PATO:0000001\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# print the content of the new schema in LinkML YAML format to view here in the notebook\n",
    "yaml_content = yaml_dumper.dumps(target_schema_obj)  # Serialize to a string\n",
    "print(yaml_content)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48d65b0bfa68ba6d",
   "metadata": {},
   "source": [
    "Notes:\n",
    "* notice that if we remove a parent class from this, e.g. remove \"NamedThing\" - the `is_a` path in the descendent classes will be absent.  this is to prevent unreachable element errors.\n",
    "* notice that helpful defaults are brought in like prefixes, descriptions, aliases, mappings, etc.\n",
    "  * there will likely be cases where metamodel elements in LinkML are not automatically transferred to the derived schema (as will all our generators, we are working towards feature parity).  \n",
    "* notice that the transformation automatically makes what were, independent `slot` definitions in Biolink Model into `attributes`.  These are more or less functionally equivalent in LinkML, however if you want to specify a slot that can be reused outside of a particular class, it is still best practice to do so with a slot definition rather than an attribute as to not repeat slot definitions."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9addb6853dbd3759",
   "metadata": {},
   "source": [
    "### Use derived schema to generate different serializations of the derived model\n",
    "\n",
    "Now that we have a YAML dump of our derived model, we can use standard LinkML tooling to produce:\n",
    "* python dataclasses and pydantic models of our derived schema\n",
    "* navigate our derived schema with a SchemaView instance\n",
    "* create and deploy automated documentation with the derived schema (see https://github.com/linkml/linkml-project-cookiecutter for more details on using the derived schema in a standard setup)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "f8ccc4af61d75af7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "from __future__ import annotations \n",
      "from datetime import (\n",
      "    datetime,\n",
      "    date\n",
      ")\n",
      "from decimal import Decimal \n",
      "from enum import Enum \n",
      "import re\n",
      "import sys\n",
      "from typing import (\n",
      "    Any,\n",
      "    ClassVar,\n",
      "    List,\n",
      "    Literal,\n",
      "    Dict,\n",
      "    Optional,\n",
      "    Union\n",
      ")\n",
      "from pydantic.version import VERSION  as PYDANTIC_VERSION \n",
      "if int(PYDANTIC_VERSION[0])>=2:\n",
      "    from pydantic import (\n",
      "        BaseModel,\n",
      "        ConfigDict,\n",
      "        Field,\n",
      "        RootModel,\n",
      "        field_validator\n",
      "    )\n",
      "else:\n",
      "    from pydantic import (\n",
      "        BaseModel,\n",
      "        Field,\n",
      "        validator\n",
      "    )\n",
      "\n",
      "metamodel_version = \"None\"\n",
      "version = \"None\"\n",
      "\n",
      "\n",
      "class ConfiguredBaseModel(BaseModel):\n",
      "    model_config = ConfigDict(\n",
      "        validate_assignment = True,\n",
      "        validate_default = True,\n",
      "        extra = \"forbid\",\n",
      "        arbitrary_types_allowed = True,\n",
      "        use_enum_values = True,\n",
      "        strict = False,\n",
      "    )\n",
      "    pass\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "class LinkMLMeta(RootModel):\n",
      "    root: Dict[str, Any] = {}\n",
      "    model_config = ConfigDict(frozen=True)\n",
      "\n",
      "    def __getattr__(self, key:str):\n",
      "        return getattr(self.root, key)\n",
      "\n",
      "    def __getitem__(self, key:str):\n",
      "        return self.root[key]\n",
      "\n",
      "    def __setitem__(self, key:str, value):\n",
      "        self.root[key] = value\n",
      "\n",
      "\n",
      "linkml_meta = LinkMLMeta({'default_prefix': 'biolink-subset/',\n",
      "     'id': 'biolink-subset',\n",
      "     'name': 'BiolinkSubset'} )\n",
      "\n",
      "\n",
      "class Gene(ConfiguredBaseModel):\n",
      "    \"\"\"\n",
      "    A region (or regions) that includes all of the sequence elements necessary to encode a functional transcript. A gene locus may include regulatory regions, transcribed regions and/or other functional sequence regions.\n",
      "    \"\"\"\n",
      "    linkml_meta: ClassVar[LinkMLMeta] = LinkMLMeta({'broad_mappings': ['NCIT:C45822'],\n",
      "         'exact_mappings': ['SO:0000704', 'SIO:010035', 'WIKIDATA:Q7187', 'dcid:Gene'],\n",
      "         'from_schema': 'biolink-subset',\n",
      "         'id_prefixes': ['NCBIGene',\n",
      "                         'ENSEMBL',\n",
      "                         'HGNC',\n",
      "                         'MGI',\n",
      "                         'ZFIN',\n",
      "                         'dictyBase',\n",
      "                         'WB',\n",
      "                         'WormBase',\n",
      "                         'FB',\n",
      "                         'RGD',\n",
      "                         'SGD',\n",
      "                         'PomBase',\n",
      "                         'OMIM',\n",
      "                         'KEGG.GENES',\n",
      "                         'UMLS',\n",
      "                         'Xenbase',\n",
      "                         'AspGD',\n",
      "                         'PHARMGKB.GENE'],\n",
      "         'in_subset': ['translator_minimal', 'model_organism_database'],\n",
      "         'narrow_mappings': ['bioschemas:gene']})\n",
      "\n",
      "    symbol: Optional[str] = Field(None, description=\"\"\"Symbol for a particular thing\"\"\", json_schema_extra = { \"linkml_meta\": {'alias': 'symbol',\n",
      "         'domain': 'named thing',\n",
      "         'domain_of': ['gene', 'Gene'],\n",
      "         'exact_mappings': ['AGRKB:symbol', 'gpi:DB_Object_Symbol'],\n",
      "         'is_a': 'node property'} })\n",
      "    xref: Optional[List[str]] = Field(default_factory=list, description=\"\"\"A database cross reference or alternative identifier for a NamedThing or edge between two NamedThings.  This property should point to a database record or webpage that supports the existence of the edge, or gives more detail about the edge. This property can be used on a node or edge to provide multiple URIs or CURIE cross references.\"\"\", json_schema_extra = { \"linkml_meta\": {'alias': 'xref',\n",
      "         'aliases': ['dbxref', 'Dbxref', 'DbXref', 'record_url', 'source_record_urls'],\n",
      "         'domain': 'named thing',\n",
      "         'domain_of': ['named thing',\n",
      "                       'publication',\n",
      "                       'retrieval source',\n",
      "                       'gene',\n",
      "                       'gene product mixin',\n",
      "                       'Gene'],\n",
      "         'in_subset': ['translator_minimal'],\n",
      "         'narrow_mappings': ['gff3:Dbxref', 'gpi:DB_Xrefs']} })\n",
      "\n",
      "\n",
      "class Disease(ConfiguredBaseModel):\n",
      "    \"\"\"\n",
      "    A disorder of structure or function, especially one that produces specific signs, phenotypes or symptoms or that affects a specific location and is not simply a direct result of physical injury.  A disposition to undergo pathological processes that exists in an organism because of one or more disorders in that organism.\n",
      "    \"\"\"\n",
      "    linkml_meta: ClassVar[LinkMLMeta] = LinkMLMeta({'aliases': ['condition', 'disorder', 'medical condition'],\n",
      "         'exact_mappings': ['MONDO:0000001',\n",
      "                            'DOID:4',\n",
      "                            'NCIT:C2991',\n",
      "                            'WIKIDATA:Q12136',\n",
      "                            'SIO:010299',\n",
      "                            'UMLSSG:DISO',\n",
      "                            'STY:T047',\n",
      "                            'dcid:Disease'],\n",
      "         'from_schema': 'biolink-subset',\n",
      "         'id_prefixes': ['MONDO',\n",
      "                         'DOID',\n",
      "                         'OMIM',\n",
      "                         'OMIM.PS',\n",
      "                         'orphanet',\n",
      "                         'EFO',\n",
      "                         'UMLS',\n",
      "                         'MESH',\n",
      "                         'MEDDRA',\n",
      "                         'NCIT',\n",
      "                         'SNOMEDCT',\n",
      "                         'medgen',\n",
      "                         'ICD10',\n",
      "                         'ICD9',\n",
      "                         'KEGG.DISEASE',\n",
      "                         'HP',\n",
      "                         'MP',\n",
      "                         'PHARMGKB.DISEASE'],\n",
      "         'in_subset': ['model_organism_database', 'translator_minimal'],\n",
      "         'narrow_mappings': ['STY:T019',\n",
      "                             'STY:T020',\n",
      "                             'STY:T048',\n",
      "                             'STY:T049',\n",
      "                             'STY:T191',\n",
      "                             'MONDO:0042489']})\n",
      "\n",
      "    pass\n",
      "\n",
      "\n",
      "class CaseToPhenotypicFeatureAssociation(ConfiguredBaseModel):\n",
      "    \"\"\"\n",
      "    An association between a case (e.g. individual patient) and a phenotypic feature in which the individual has or has had the phenotype.\n",
      "    \"\"\"\n",
      "    linkml_meta: ClassVar[LinkMLMeta] = LinkMLMeta({'defining_slots': ['subject', 'object'], 'from_schema': 'biolink-subset'})\n",
      "\n",
      "    pass\n",
      "\n",
      "\n",
      "class GeneToDiseaseAssociation(ConfiguredBaseModel):\n",
      "    linkml_meta: ClassVar[LinkMLMeta] = LinkMLMeta({'close_mappings': ['dcid:DiseaseGeneAssociation'],\n",
      "         'comments': ['NCIT:R176 refers to the inverse relationship'],\n",
      "         'defining_slots': ['subject', 'object'],\n",
      "         'exact_mappings': ['SIO:000983'],\n",
      "         'from_schema': 'biolink-subset'})\n",
      "\n",
      "    pass\n",
      "\n",
      "\n",
      "class GeneToPhenotypicFeatureAssociation(ConfiguredBaseModel):\n",
      "    linkml_meta: ClassVar[LinkMLMeta] = LinkMLMeta({'defining_slots': ['subject', 'object'],\n",
      "         'exact_mappings': ['WBVocab:Gene-Phenotype-Association'],\n",
      "         'from_schema': 'biolink-subset'})\n",
      "\n",
      "    pass\n",
      "\n",
      "\n",
      "class Case(ConfiguredBaseModel):\n",
      "    \"\"\"\n",
      "    An individual (human) organism that has a patient role in some clinical context.\n",
      "    \"\"\"\n",
      "    linkml_meta: ClassVar[LinkMLMeta] = LinkMLMeta({'aliases': ['patient', 'proband'], 'from_schema': 'biolink-subset'})\n",
      "\n",
      "    pass\n",
      "\n",
      "\n",
      "class PhenotypicFeature(ConfiguredBaseModel):\n",
      "    \"\"\"\n",
      "    A combination of entity and quality that makes up a phenotyping statement. An observable characteristic of an individual resulting from the interaction of its genotype with its molecular and physical environment.\n",
      "    \"\"\"\n",
      "    linkml_meta: ClassVar[LinkMLMeta] = LinkMLMeta({'aliases': ['sign', 'symptom', 'phenotype', 'trait', 'endophenotype'],\n",
      "         'broad_mappings': ['BFO:0000019', 'PATO:0000001'],\n",
      "         'exact_mappings': ['UPHENO:0001001',\n",
      "                            'SIO:010056',\n",
      "                            'WIKIDATA:Q104053',\n",
      "                            'UMLS:C4021819',\n",
      "                            'NCIT:C16977',\n",
      "                            'SNOMEDCT:8116006',\n",
      "                            'MESH:D010641'],\n",
      "         'examples': [{'description': 'decreased body weight', 'value': 'MP:0001262'}],\n",
      "         'from_schema': 'biolink-subset',\n",
      "         'id_prefixes': ['HP',\n",
      "                         'EFO',\n",
      "                         'NCIT',\n",
      "                         'UMLS',\n",
      "                         'MEDDRA',\n",
      "                         'MP',\n",
      "                         'ZP',\n",
      "                         'UPHENO',\n",
      "                         'APO',\n",
      "                         'FBcv',\n",
      "                         'WBPhenotype',\n",
      "                         'SNOMEDCT',\n",
      "                         'MESH',\n",
      "                         'XPO',\n",
      "                         'FYPO',\n",
      "                         'TO'],\n",
      "         'in_subset': ['model_organism_database'],\n",
      "         'narrow_mappings': ['STY:T184',\n",
      "                             'WIKIDATA:Q169872',\n",
      "                             'WIKIDATA:Q25203551',\n",
      "                             'ZP:00000000',\n",
      "                             'FBcv:0001347',\n",
      "                             'HP:0000118',\n",
      "                             'MP:0000001',\n",
      "                             'WBPhenotype:0000886',\n",
      "                             'XPO:00000000',\n",
      "                             'FYPO:0000001',\n",
      "                             'APO:0000017',\n",
      "                             'TO:0000387',\n",
      "                             'STY:T190']})\n",
      "\n",
      "    pass\n",
      "\n",
      "\n",
      "# Model rebuild\n",
      "# see https://pydantic-docs.helpmanual.io/usage/models/#rebuilding-a-model\n",
      "Gene.model_rebuild()\n",
      "Disease.model_rebuild()\n",
      "CaseToPhenotypicFeatureAssociation.model_rebuild()\n",
      "GeneToDiseaseAssociation.model_rebuild()\n",
      "GeneToPhenotypicFeatureAssociation.model_rebuild()\n",
      "Case.model_rebuild()\n",
      "PhenotypicFeature.model_rebuild()\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!gen-pydantic biolink-subset.yaml"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "3c811e7ae6de6e3a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ValueError: No such slot as \"node property\"\n"
     ]
    }
   ],
   "source": [
    "!gen-doc biolink-subset.yaml -d docs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5ef1128-c499-4359-abea-7cff4056d761",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
